### PR TITLE
`build/pkgs/ccache`: Update to 4.10.2

### PR DESCRIPTION
**Pull request by gmou3**
I also modified `sdh_cmake` slightly, because inputting `.` and `..` together issued a warning.

Resolves #35872.

### CODE REVIEW

File: src/doc/en/developer/packaging.rst [hunk 1]
```diff
@@ -497,11 +497,10 @@ should not need to add it yourself.
...
 
-- ``sdh_cmake [...]``: Runs ``cmake`` in the current directory with
-  the given arguments, as well as additional arguments passed to
-  cmake (assuming packages are using the GNUInstallDirs module) so
-  that ``CMAKE_INSTALL_PREFIX`` and ``CMAKE_INSTALL_LIBDIR`` are set
-  correctly.
+- ``sdh_cmake [...]``: Runs ``cmake`` in the current directory with the given
+  arguments, as well as additional arguments (assuming packages are using the
+  GNUInstallDirs module) so that ``CMAKE_INSTALL_PREFIX`` and
+  ``CMAKE_INSTALL_LIBDIR`` are set correctly.
```
Comment (kwankyu) [inline src/doc/en/developer/packaging.rst:504]: Because of your change, `in the current directory` should be removed?

Comment (gmou3) [reply src/doc/en/developer/packaging.rst:504]: Indeed, I somehow missed that.

**Comment by kwankyu** [comment]
I tested the branch (and the new ccache) on a mac. Did you test it on any linux?

**Comment by gmou3** [comment]
Yes, seems to work fine.

**Review by kwankyu — APPROVED** [review]
OK. LGTM.

Thanks.

**Comment by enriqueartal** [comment]
I had an issue in Fedora 41, `ccache` installation did not work having some issue with `Xxhash`. It worked after installing system package xxhash-devel; should this dependency appeared somewhere? On the other side system package ccache is installed and a `fedora.txt` exists in the corresponding distro folder, but it is ignored.

**Comment by kwankyu** [comment]
See https://groups.google.com/g/sage-devel/c/ZZIvoEHm5hg for the issue.

**Comment by dimpase** [comment]
Why does this PR touch a lot a files not related in any way to ccache?

**Comment by dimpase** [comment]
Another problematic change is that ccache now needs xxhash, and it tries to download it from the internet - something that probably breaks the rules for optional packages. And it also is flaky, see
https://groups.google.com/d/msgid/sage-release/CAD0p0K51VNkbovm1De3arXt18E9rpoP%2B%3DWO1414Cb-b_FP0%3D6Q

**Comment by gmou3** [comment]
> Why does this PR touch a lot a files not related in any way to ccache?

I also modified `sdh_cmake` to not automatically run `cmake` in the current directory. This allows us to follow a package's installation instructions as recommended by the author (e.g., possibly `cmake ..`).

> Another problematic change is that ccache now needs xxhash, and it tries to download it from the internet - something that probably breaks the rules for optional packages. And it also is flaky, see https://groups.google.com/d/msgid/sage-release/CAD0p0K51VNkbovm1De3arXt18E9rpoP%2B%3DWO1414Cb-b_FP0%3D6Q%40mail.gmail.com

Indeed, this is an issue. See https://github.com/sagemath/sage/pull/39226 for the proposed solution.

**Comment by gmou3** [comment]
That's unexpected, its `optional` status hasn't changed.
I'll try a clean build on my machine later to see if I can reproduce the issue (did you try this by any chance?).
I can't post on the discussion you shared, but it seems that one of the two users has a problem relevant to `ccache` while the other does not.

**Comment by dimpase** [comment]
> That's unexpected, its `optional` status hasn't changed. I'll try a clean build on my machine later to see if I can reproduce the issue (did you try this by any chance?). I can't post on the discussion you shared, but it seems that one of the two users has a problem relevant to `ccache` while the other does not.

no, at least 2 users, @JohnCremona and @EmmanuelCharpentier  report unwanted installation of ccache. It could be they run same or similar versions of Ubuntu.

**Comment by EmmanuelCharpentier** [comment]
Le jeudi 23 janvier 2025 à 10:58 -0800, Dima Pasechnik a écrit :
> > That's unexpected, its optional status hasn't changed. I'll try a
> > clean build on my machine later to see if I can reproduce the issue
> > (did you try this by any chance?). I can't post on the discussion
> > you shared, but it seems that one of the two users has a problem
> > relevant to ccache while the other does not.
> no, at least 2 users, @JohnCremona [1] and @EmmanuelCharpentier [2]
> report unwanted installation of ccache. It could be they run same or
> similar versions of Ubuntu.

I'm running Debian testing, upgraded (more or less) daily.

--
Emmanuel Charpentier

> —
> Reply to this email directly, view it on GitHub [3], or unsubscribe
> [4].
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>


[1] @JohnCremona https://github.com/JohnCremona
[2] @EmmanuelCharpentier https://github.com/EmmanuelCharpentier
[3] view it on GitHub
    https://github.com/sagemath/sage/pull/39119#issuecomment-2610779917
[4] unsubscribe

**Comment by JohnCremona** [comment]
My machine runs Ubuntu 22

**Comment by dimpase** [comment]
> > Why does this PR touch a lot a files not related in any way to ccache?
> 
> I also modified `sdh_cmake` to not automatically run `cmake` in the current directory. This allows us to follow a package's installation instructions as recommended by the author (e.g., possibly `cmake ..`).

I don't think it's needed any more. Early versions of cmake required a separate build directory, and these instructions might be a leftover from these early days.

Did you try without making these changes?

------------------

I don't see this problem (of unrequested building of ccache spkg) on Gentoo, so this makes for an "interesting" platform-specific bug...

**Comment by gmou3** [comment]
> > > Why does this PR touch a lot a files not related in any way to ccache?
> > 
> > 
> > I also modified `sdh_cmake` to not automatically run `cmake` in the current directory. This allows us to follow a package's installation instructions as recommended by the author (e.g., possibly `cmake ..`).
> 
> I don't think it's needed any more. Early versions of cmake required a separate build directory, and these instructions might be a leftover from these early days.
> 
> Did you try without making these changes?
>
`cmake` requires a build directory, and often `..` is used. Not only `.`.

Not making such changes would require me to diverge from the recommended `ccache` instructions, by I wanted to avoid the uncertainties of this. I think this makes the translation of build instructions easier and clearer for all packages. `cmake` becomes `sdh_cmake` and that's it.

> I don't see this problem (of unrequested building of ccache spkg) on Gentoo, so this makes for an "interesting" platform-specific bug...

So it probably isn't a result of this PR. Maybe the failure just made the building of `ccache` apparent.

**Comment by dimpase** [comment]
> So it probably isn't a result of this PR. Maybe the failure just made the building of `ccache` apparent.

@JohnCremona - could you please check whether reverting this PR (i.e. the commits at https://github.com/sagemath/sage/pull/39119/commits) fixed the issue of the unasked for build of `ccache` ?

**Comment by JohnCremona** [comment]
> > So it probably isn't a result of this PR. Maybe the failure just made the building of `ccache` apparent.
> 
> @JohnCremona - could you please check whether reverting this PR (i.e. the commits at https://github.com/sagemath/sage/pull/39119/commits) fixed the issue of the unasked for build of `ccache` ?

Sorry for the deelay, I was away.  I will try.

**Pull request merged**